Tools/TestWebKitAPI/mac/DragAndDropSimulatorMac.mm

```diff
@@ -151,6 +151,23 @@ - (double)initialProgressForMouseDrag
     return !totalDistance ? 1 : std::min<double>(1, initialMouseDragDistance / totalDistance);
 }
 
+- (void)runFromElement:(NSString *)startID toElement:(NSString *)endID
```

Comment:
We should probably also put this implementation in a single place, instead of copying it in two places — e.g. `DragAndDropSimulator.mm`. You can probably use an anonymous category:

```
@implementation DragAndDropSimulator ()
…
@end
```
...for that.

Reply:
I've put the implementation in a single place (`cocoa/DragAndDropSimulator.mm`), but couldn't get the anonymous category approach to work. What do you think of the current solution in https://github.com/WebKit/WebKit/pull/13844/commits/e92cb7bf8985c7c8b09a62d159bf24619ec278dd? And/or do you know why this didn't work?

```obj-c
// In DragAndDropSimulator.mm

@interface DragAndDropSimulator ()

- (void)runFromElement:(NSString *)startSelector toElement:(NSString *)endSelector;

@end


@implementation DragAndDropSimulator

- (void)runFromElement:(NSString *)startSelector toElement:(NSString *)endSelector
{
    // ...
}

@end
```

The compiler suggests `@implementation DragAndDropSimulator` in `DragAndDropSimulator.mm` should implement the full interface `@interface DragAndDropSimulator : NSObject<WKUIDelegatePrivate, _WKInputDelegate>` described in `DragAndDropSimulator.h`.

<details>
<summary>Build error from the anonymous category approach</summary>

```
TestWebKitAPILibrary CompileC DragAndDropTestsIOS.o
/Users/aprotyas/dev/safari/OpenSource/Tools/TestWebKitAPI/Tests/ios/DragAndDropTestsIOS.mm:1214:20: error: instance method '-runFromElement:toElement:' not found (return type defaults to 'id') [-Werror,-Wobjc-method-access]
        [simulator runFromElement:@"#output" toElement:@"#droparea"];
In file included from /Users/aprotyas/dev/safari/OpenSource/Tools/TestWebKitAPI/Tests/ios/DragAndDropTestsIOS.mm:31:
/Users/aprotyas/dev/safari/OpenSource/Tools/TestWebKitAPI/cocoa/DragAndDropSimulator.h:79:12: note: receiver is instance of class declared here
@interface DragAndDropSimulator : NSObject<WKUIDelegatePrivate, _WKInputDelegate>
/Users/aprotyas/dev/safari/OpenSource/Tools/TestWebKitAPI/Tests/ios/DragAndDropTestsIOS.mm:1244:20: error: instance method '-runFromElement:toElement:' not found (return type defaults to 'id') [-Werror,-Wobjc-method-access]
        [simulator runFromElement:@"#output" toElement:@"#droparea"];
In file included from /Users/aprotyas/dev/safari/OpenSource/Tools/TestWebKitAPI/Tests/ios/DragAndDropTestsIOS.mm:31:
/Users/aprotyas/dev/safari/OpenSource/Tools/TestWebKitAPI/cocoa/DragAndDropSimulator.h:79:12: note: receiver is instance of class declared here
@interface DragAndDropSimulator : NSObject<WKUIDelegatePrivate, _WKInputDelegate>
TestWebKitAPILibrary CompileC UnifiedSource9-mm.o
TestWebKitAPILibrary CompileC UnifiedSource39-mm.o
In file included from /Users/aprotyas/Build/Products/Debug-iphonesimulator/DerivedSources/TestWebKitAPI/unified-sources/UnifiedSource39-mm.mm:2:
/Users/aprotyas/dev/safari/OpenSource/Tools/TestWebKitAPI/cocoa/DragAndDropSimulator.mm:41:17: error: method definition for 'initWithWebViewFrame:' not found [-Werror,-Wincomplete-implementation]
@implementation DragAndDropSimulator
In file included from /Users/aprotyas/Build/Products/Debug-iphonesimulator/DerivedSources/TestWebKitAPI/unified-sources/UnifiedSource39-mm.mm:2:
In file included from /Users/aprotyas/dev/safari/OpenSource/Tools/TestWebKitAPI/cocoa/DragAndDropSimulator.mm:27:
/Users/aprotyas/dev/safari/OpenSource/Tools/TestWebKitAPI/cocoa/DragAndDropSimulator.h:81:1: note: method 'initWithWebViewFrame:' declared here
- (instancetype)initWithWebViewFrame:(CGRect)frame;
^
In file included from /Users/aprotyas/Build/Products/Debug-iphonesimulator/DerivedSources/TestWebKitAPI/unified-sources/UnifiedSource39-mm.mm:2:
/Users/aprotyas/dev/safari/OpenSource/Tools/TestWebKitAPI/cocoa/DragAndDropSimulator.mm:41:17: error: method definition for 'initWithWebViewFrame:configuration:' not found [-Werror,-Wincomplete-implementation]
@implementation DragAndDropSimulator
In file included from /Users/aprotyas/Build/Products/Debug-iphonesimulator/DerivedSources/TestWebKitAPI/unified-sources/UnifiedSource39-mm.mm:2:
In file included from /Users/aprotyas/dev/safari/OpenSource/Tools/TestWebKitAPI/cocoa/DragAndDropSimulator.mm:27:
/Users/aprotyas/dev/safari/OpenSource/Tools/TestWebKitAPI/cocoa/DragAndDropSimulator.h:82:1: note: method 'initWithWebViewFrame:configuration:' declared here
- (instancetype)initWithWebViewFrame:(CGRect)frame configuration:(WKWebViewConfiguration *)configuration;
^
In file included from /Users/aprotyas/Build/Products/Debug-iphonesimulator/DerivedSources/TestWebKitAPI/unified-sources/UnifiedSource39-mm.mm:2:
/Users/aprotyas/dev/safari/OpenSource/Tools/TestWebKitAPI/cocoa/DragAndDropSimulator.mm:41:17: error: method definition for 'runFrom:to:' not found [-Werror,-Wincomplete-implementation]
@implementation DragAndDropSimulator
In file included from /Users/aprotyas/Build/Products/Debug-iphonesimulator/DerivedSources/TestWebKitAPI/unified-sources/UnifiedSource39-mm.mm:2:
In file included from /Users/aprotyas/dev/safari/OpenSource/Tools/TestWebKitAPI/cocoa/DragAndDropSimulator.mm:27:
/Users/aprotyas/dev/safari/OpenSource/Tools/TestWebKitAPI/cocoa/DragAndDropSimulator.h:84:1: note: method 'runFrom:to:' declared here
- (void)runFrom:(CGPoint)startLocation to:(CGPoint)endLocation;
^
In file included from /Users/aprotyas/Build/Products/Debug-iphonesimulator/DerivedSources/TestWebKitAPI/unified-sources/UnifiedSource39-mm.mm:2:
/Users/aprotyas/dev/safari/OpenSource/Tools/TestWebKitAPI/cocoa/DragAndDropSimulator.mm:41:17: error: method definition for 'endDataTransfer' not found [-Werror,-Wincomplete-implementation]
@implementation DragAndDropSimulator
In file included from /Users/aprotyas/Build/Products/Debug-iphonesimulator/DerivedSources/TestWebKitAPI/unified-sources/UnifiedSource39-mm.mm:2:
In file included from /Users/aprotyas/dev/safari/OpenSource/Tools/TestWebKitAPI/cocoa/DragAndDropSimulator.mm:27:
/Users/aprotyas/dev/safari/OpenSource/Tools/TestWebKitAPI/cocoa/DragAndDropSimulator.h:85:1: note: method 'endDataTransfer' declared here
- (void)endDataTransfer;
^
In file included from /Users/aprotyas/Build/Products/Debug-iphonesimulator/DerivedSources/TestWebKitAPI/unified-sources/UnifiedSource39-mm.mm:2:
/Users/aprotyas/dev/safari/OpenSource/Tools/TestWebKitAPI/cocoa/DragAndDropSimulator.mm:41:17: error: method definition for 'clearExternalDragInformation' not found [-Werror,-Wincomplete-implementation]
@implementation DragAndDropSimulator
In file included from /Users/aprotyas/Build/Products/Debug-iphonesimulator/DerivedSources/TestWebKitAPI/unified-sources/UnifiedSource39-mm.mm:2:
In file included from /Users/aprotyas/dev/safari/OpenSource/Tools/TestWebKitAPI/cocoa/DragAndDropSimulator.mm:27:
/Users/aprotyas/dev/safari/OpenSource/Tools/TestWebKitAPI/cocoa/DragAndDropSimulator.h:86:1: note: method 'clearExternalDragInformation' declared here
- (void)clearExternalDragInformation;
^
In file included from /Users/aprotyas/Build/Products/Debug-iphonesimulator/DerivedSources/TestWebKitAPI/unified-sources/UnifiedSource39-mm.mm:2:
/Users/aprotyas/dev/safari/OpenSource/Tools/TestWebKitAPI/cocoa/DragAndDropSimulator.mm:41:17: error: method definition for 'initWithWebView:' not found [-Werror,-Wincomplete-implementation]
@implementation DragAndDropSimulator
In file included from /Users/aprotyas/Build/Products/Debug-iphonesimulator/DerivedSources/TestWebKitAPI/unified-sources/UnifiedSource39-mm.mm:2:
In file included from /Users/aprotyas/dev/safari/OpenSource/Tools/TestWebKitAPI/cocoa/DragAndDropSimulator.mm:27:
/Users/aprotyas/dev/safari/OpenSource/Tools/TestWebKitAPI/cocoa/DragAndDropSimulator.h:94:1: note: method 'initWithWebView:' declared here
- (instancetype)initWithWebView:(TestWKWebView *)webView;
^
In file included from /Users/aprotyas/Build/Products/Debug-iphonesimulator/DerivedSources/TestWebKitAPI/unified-sources/UnifiedSource39-mm.mm:2:
/Users/aprotyas/dev/safari/OpenSource/Tools/TestWebKitAPI/cocoa/DragAndDropSimulator.mm:41:17: error: method definition for 'runFrom:to:additionalItemRequestLocations:' not found [-Werror,-Wincomplete-implementation]
@implementation DragAndDropSimulator
In file included from /Users/aprotyas/Build/Products/Debug-iphonesimulator/DerivedSources/TestWebKitAPI/unified-sources/UnifiedSource39-mm.mm:2:
In file included from /Users/aprotyas/dev/safari/OpenSource/Tools/TestWebKitAPI/cocoa/DragAndDropSimulator.mm:27:
/Users/aprotyas/dev/safari/OpenSource/Tools/TestWebKitAPI/cocoa/DragAndDropSimulator.h:95:1: note: method 'runFrom:to:additionalItemRequestLocations:' declared here
- (void)runFrom:(CGPoint)startLocation to:(CGPoint)endLocation additionalItemRequestLocations:(ProgressToCGPointValueMap)additionalItemRequestLocations;
^
In file included from /Users/aprotyas/Build/Products/Debug-iphonesimulator/DerivedSources/TestWebKitAPI/unified-sources/UnifiedSource39-mm.mm:2:
/Users/aprotyas/dev/safari/OpenSource/Tools/TestWebKitAPI/cocoa/DragAndDropSimulator.mm:41:17: error: method definition for 'ensureInputSession' not found [-Werror,-Wincomplete-implementation]
@implementation DragAndDropSimulator
In file included from /Users/aprotyas/Build/Products/Debug-iphonesimulator/DerivedSources/TestWebKitAPI/unified-sources/UnifiedSource39-mm.mm:2:
In file included from /Users/aprotyas/dev/safari/OpenSource/Tools/TestWebKitAPI/cocoa/DragAndDropSimulator.mm:27:
/Users/aprotyas/dev/safari/OpenSource/Tools/TestWebKitAPI/cocoa/DragAndDropSimulator.h:96:1: note: method 'ensureInputSession' declared here
- (void)ensureInputSession;
^
In file included from /Users/aprotyas/Build/Products/Debug-iphonesimulator/DerivedSources/TestWebKitAPI/unified-sources/UnifiedSource39-mm.mm:2:
/Users/aprotyas/dev/safari/OpenSource/Tools/TestWebKitAPI/cocoa/DragAndDropSimulator.mm:41:17: error: method definition for 'setExternalItemProviders:defaultDropPreviews:' not found [-Werror,-Wincomplete-implementation]
@implementation DragAndDropSimulator
In file included from /Users/aprotyas/Build/Products/Debug-iphonesimulator/DerivedSources/TestWebKitAPI/unified-sources/UnifiedSource39-mm.mm:2:
In file included from /Users/aprotyas/dev/safari/OpenSource/Tools/TestWebKitAPI/cocoa/DragAndDropSimulator.mm:27:
/Users/aprotyas/dev/safari/OpenSource/Tools/TestWebKitAPI/cocoa/DragAndDropSimulator.h:97:1: note: method 'setExternalItemProviders:defaultDropPreviews:' declared here
- (void)setExternalItemProviders:(NSArray<NSItemProvider *> *)itemProviders defaultDropPreviews:(NSArray<UITargetedDragPreview *> *)previews;
^
In file included from /Users/aprotyas/Build/Products/Debug-iphonesimulator/DerivedSources/TestWebKitAPI/unified-sources/UnifiedSource39-mm.mm:2:
/Users/aprotyas/dev/safari/OpenSource/Tools/TestWebKitAPI/cocoa/DragAndDropSimulator.mm:41:17: error: method definition for 'containsDraggedType:' not found [-Werror,-Wincomplete-implementation]
@implementation DragAndDropSimulator
In file included from /Users/aprotyas/Build/Products/Debug-iphonesimulator/DerivedSources/TestWebKitAPI/unified-sources/UnifiedSource39-mm.mm:2:
In file included from /Users/aprotyas/dev/safari/OpenSource/Tools/TestWebKitAPI/cocoa/DragAndDropSimulator.mm:27:
/Users/aprotyas/dev/safari/OpenSource/Tools/TestWebKitAPI/cocoa/DragAndDropSimulator.h:145:1: note: method 'containsDraggedType:' declared here
- (BOOL)containsDraggedType:(NSString *)type;
^
TestWebKitAPILibrary CompileC DragAndDropSimulator.o
/Users/aprotyas/dev/safari/OpenSource/Tools/TestWebKitAPI/cocoa/DragAndDropSimulator.mm:41:17: error: method definition for 'initWithWebViewFrame:' not found [-Werror,-Wincomplete-implementation]
@implementation DragAndDropSimulator
In file included from /Users/aprotyas/dev/safari/OpenSource/Tools/TestWebKitAPI/cocoa/DragAndDropSimulator.mm:27:
/Users/aprotyas/dev/safari/OpenSource/Tools/TestWebKitAPI/cocoa/DragAndDropSimulator.h:81:1: note: method 'initWithWebViewFrame:' declared here
- (instancetype)initWithWebViewFrame:(CGRect)frame;
^
/Users/aprotyas/dev/safari/OpenSource/Tools/TestWebKitAPI/cocoa/DragAndDropSimulator.mm:41:17: error: method definition for 'initWithWebViewFrame:configuration:' not found [-Werror,-Wincomplete-implementation]
@implementation DragAndDropSimulator
In file included from /Users/aprotyas/dev/safari/OpenSource/Tools/TestWebKitAPI/cocoa/DragAndDropSimulator.mm:27:
/Users/aprotyas/dev/safari/OpenSource/Tools/TestWebKitAPI/cocoa/DragAndDropSimulator.h:82:1: note: method 'initWithWebViewFrame:configuration:' declared here
- (instancetype)initWithWebViewFrame:(CGRect)frame configuration:(WKWebViewConfiguration *)configuration;
^
/Users/aprotyas/dev/safari/OpenSource/Tools/TestWebKitAPI/cocoa/DragAndDropSimulator.mm:41:17: error: method definition for 'runFrom:to:' not found [-Werror,-Wincomplete-implementation]
@implementation DragAndDropSimulator
In file included from /Users/aprotyas/dev/safari/OpenSource/Tools/TestWebKitAPI/cocoa/DragAndDropSimulator.mm:27:
/Users/aprotyas/dev/safari/OpenSource/Tools/TestWebKitAPI/cocoa/DragAndDropSimulator.h:84:1: note: method 'runFrom:to:' declared here
- (void)runFrom:(CGPoint)startLocation to:(CGPoint)endLocation;
^
/Users/aprotyas/dev/safari/OpenSource/Tools/TestWebKitAPI/cocoa/DragAndDropSimulator.mm:41:17: error: method definition for 'endDataTransfer' not found [-Werror,-Wincomplete-implementation]
@implementation DragAndDropSimulator
In file included from /Users/aprotyas/dev/safari/OpenSource/Tools/TestWebKitAPI/cocoa/DragAndDropSimulator.mm:27:
/Users/aprotyas/dev/safari/OpenSource/Tools/TestWebKitAPI/cocoa/DragAndDropSimulator.h:85:1: note: method 'endDataTransfer' declared here
- (void)endDataTransfer;
^
/Users/aprotyas/dev/safari/OpenSource/Tools/TestWebKitAPI/cocoa/DragAndDropSimulator.mm:41:17: error: method definition for 'clearExternalDragInformation' not found [-Werror,-Wincomplete-implementation]
@implementation DragAndDropSimulator
In file included from /Users/aprotyas/dev/safari/OpenSource/Tools/TestWebKitAPI/cocoa/DragAndDropSimulator.mm:27:
/Users/aprotyas/dev/safari/OpenSource/Tools/TestWebKitAPI/cocoa/DragAndDropSimulator.h:86:1: note: method 'clearExternalDragInformation' declared here
- (void)clearExternalDragInformation;
^
/Users/aprotyas/dev/safari/OpenSource/Tools/TestWebKitAPI/cocoa/DragAndDropSimulator.mm:41:17: error: method definition for 'initWithWebView:' not found [-Werror,-Wincomplete-implementation]
@implementation DragAndDropSimulator
In file included from /Users/aprotyas/dev/safari/OpenSource/Tools/TestWebKitAPI/cocoa/DragAndDropSimulator.mm:27:
/Users/aprotyas/dev/safari/OpenSource/Tools/TestWebKitAPI/cocoa/DragAndDropSimulator.h:94:1: note: method 'initWithWebView:' declared here
- (instancetype)initWithWebView:(TestWKWebView *)webView;
^
/Users/aprotyas/dev/safari/OpenSource/Tools/TestWebKitAPI/cocoa/DragAndDropSimulator.mm:41:17: error: method definition for 'runFrom:to:additionalItemRequestLocations:' not found [-Werror,-Wincomplete-implementation]
@implementation DragAndDropSimulator
In file included from /Users/aprotyas/dev/safari/OpenSource/Tools/TestWebKitAPI/cocoa/DragAndDropSimulator.mm:27:
/Users/aprotyas/dev/safari/OpenSource/Tools/TestWebKitAPI/cocoa/DragAndDropSimulator.h:95:1: note: method 'runFrom:to:additionalItemRequestLocations:' declared here
- (void)runFrom:(CGPoint)startLocation to:(CGPoint)endLocation additionalItemRequestLocations:(ProgressToCGPointValueMap)additionalItemRequestLocations;
^
/Users/aprotyas/dev/safari/OpenSource/Tools/TestWebKitAPI/cocoa/DragAndDropSimulator.mm:41:17: error: method definition for 'ensureInputSession' not found [-Werror,-Wincomplete-implementation]
@implementation DragAndDropSimulator
In file included from /Users/aprotyas/dev/safari/OpenSource/Tools/TestWebKitAPI/cocoa/DragAndDropSimulator.mm:27:
/Users/aprotyas/dev/safari/OpenSource/Tools/TestWebKitAPI/cocoa/DragAndDropSimulator.h:96:1: note: method 'ensureInputSession' declared here
- (void)ensureInputSession;
^
/Users/aprotyas/dev/safari/OpenSource/Tools/TestWebKitAPI/cocoa/DragAndDropSimulator.mm:41:17: error: method definition for 'setExternalItemProviders:defaultDropPreviews:' not found [-Werror,-Wincomplete-implementation]
@implementation DragAndDropSimulator
In file included from /Users/aprotyas/dev/safari/OpenSource/Tools/TestWebKitAPI/cocoa/DragAndDropSimulator.mm:27:
/Users/aprotyas/dev/safari/OpenSource/Tools/TestWebKitAPI/cocoa/DragAndDropSimulator.h:97:1: note: method 'setExternalItemProviders:defaultDropPreviews:' declared here
- (void)setExternalItemProviders:(NSArray<NSItemProvider *> *)itemProviders defaultDropPreviews:(NSArray<UITargetedDragPreview *> *)previews;
^
/Users/aprotyas/dev/safari/OpenSource/Tools/TestWebKitAPI/cocoa/DragAndDropSimulator.mm:41:17: error: method definition for 'containsDraggedType:' not found [-Werror,-Wincomplete-implementation]
@implementation DragAndDropSimulator
In file included from /Users/aprotyas/dev/safari/OpenSource/Tools/TestWebKitAPI/cocoa/DragAndDropSimulator.mm:27:
/Users/aprotyas/dev/safari/OpenSource/Tools/TestWebKitAPI/cocoa/DragAndDropSimulator.h:145:1: note: method 'containsDraggedType:' declared here
- (BOOL)containsDraggedType:(NSString *)type;
^
10 errors generated.
** BUILD FAILED **
```

</details>